testsuite/cluster-test/src/cluster_swarm/cluster_swarm_kube.rs

```diff
@@ -0,0 +1,480 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::{bail, format_err, Result};
+use async_trait::async_trait;
+
+use futures::lock::Mutex;
+use kube::{
+    api::{Api, PostParams},
+    client::APIClient,
+    config,
+};
+use slog_scope::*;
+
+use util::retry;
+
+use crate::cluster_swarm::ClusterSwarm;
+use crate::instance::Instance;
+use libra_config::config::AdmissionControlConfig;
+
+const DEFAULT_NAMESPACE: &str = "default";
+
+const ERROR_NOT_FOUND: u16 = 404;
+
+pub struct ClusterSwarmKube {
+    client: APIClient,
+    validator_to_node: Arc<Mutex<HashMap<u32, Instance>>>,
+    fullnode_to_node: Arc<Mutex<HashMap<(u32, u32), Instance>>>,
+}
+
+impl ClusterSwarmKube {
+    pub async fn new() -> Result<Self> {
+        let mut config = config::load_kube_config().await;
+        if config.is_err() {
+            config = config::incluster_config();
+        }
+        let config = config.map_err(|e| format_err!("Failed to load config: {:?}", e))?;
+        let client = APIClient::new(config);
+        let validator_to_node = Arc::new(Mutex::new(HashMap::new()));
+        let fullnode_to_node = Arc::new(Mutex::new(HashMap::new()));
+        Ok(Self {
+            client,
+            validator_to_node,
+            fullnode_to_node,
+        })
+    }
+
+    fn validator_spec(
+        &self,
+        index: u32,
+        num_validators: u32,
+        num_fullnodes: u32,
+        node_name: &str,
+        image_tag: &str,
+        delete_data: bool,
+    ) -> Vec<u8> {
+        let pod_yaml = format!(
```

Comment:
Consider putting this in separate file and using [include_str!](https://doc.rust-lang.org/std/macro.include_str.html)